wrappers/src/table_cs.cpp

```diff
@@ -28,144 +28,267 @@
 #include "timestamp_helpers.hpp"
 #include "object-store/src/results.hpp"
 #include "marshalable_sort_clause.hpp"
+#include "object_accessor.hpp"
+#include "schema.hpp"
 
 using namespace realm;
 using namespace realm::binding;
 
 
 extern "C" {
 
-REALM_EXPORT void table_unbind(Table* table_ptr, NativeException::Marshallable& ex)
+REALM_EXPORT void table_unbind(const Table* table_ptr, NativeException::Marshallable& ex)
 {
     return handle_errors(ex, [&]() {
         LangBindHelper::unbind_table_ptr(table_ptr);
     });
 }
 
-REALM_EXPORT Row* table_add_empty_row(Table* table_ptr, NativeException::Marshallable& ex)
+REALM_EXPORT Object* table_add_empty_row(Table* table_ptr, SharedRealm* realm, NativeException::Marshallable& ex)
 {
     return handle_errors(ex, [&]() {
+        realm->get()->verify_in_write();
+        
         size_t row_ndx = table_ptr->add_empty_row(1);
-        return new Row((*table_ptr)[row_ndx]);
+        const StringData object_name = ObjectStore::object_type_for_table_name(table_ptr->get_name());
+        auto object_schema = realm->get()->schema().find(object_name);
+        return new Object(*realm, *object_schema, *new Row((*table_ptr)[row_ndx]));
     });
 }
 
-REALM_EXPORT Row* table_get_link(Table* table_ptr, size_t column_ndx, size_t row_ndx, NativeException::Marshallable& ex)
+REALM_EXPORT Object* table_get_link(const Object* object_ptr, size_t property_ndx, NativeException::Marshallable& ex)
 {
-  return handle_errors(ex, [&]() -> Row* {
-    const size_t link_row_ndx = table_ptr->get_link(column_ndx, row_ndx);
-    if (link_row_ndx == realm::npos)
-      return nullptr;
-    auto target_table_ptr = table_ptr->get_link_target(column_ndx);
-    return new Row((*target_table_ptr)[link_row_ndx]);
-  });
+    return handle_errors(ex, [&]() -> Object* {
+        if (object_ptr->realm()->is_closed())
+            throw RealmClosedException();
+        
+        if (!object_ptr->is_valid())
+            throw RowDetachedException();
+
+        object_ptr->realm()->verify_thread();
+        
+        const size_t column_ndx = object_ptr->get_object_schema().persisted_properties[property_ndx].table_column;
+        const size_t link_row_ndx = object_ptr->row().get_link(column_ndx);
+        if (link_row_ndx == realm::npos)
+            return nullptr;
+        
+        auto target_table_ptr = object_ptr->row().get_table()->get_link_target(column_ndx);
+        const StringData target_name = ObjectStore::object_type_for_table_name(target_table_ptr->get_name());
+        const ObjectSchema& target_schema = *object_ptr->realm()->schema().find(target_name);
```

Comment:
this is inconsistent with e.g. https://github.com/realm/realm-dotnet/pull/866/files#diff-f651c2b696cbccfb7b7a3b23bf46e21cR54 where we use `auto target_schema = ...`